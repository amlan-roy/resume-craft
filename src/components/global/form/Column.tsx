import React from "react";
import {
  SortableContext,
  verticalListSortingStrategy,
} from "@dnd-kit/sortable";
import Task from "./Task";

type ColumnProps = {
  tasks: { id: Number; title: String }[];
};

const Column: React.FC<ColumnProps> = ({ tasks }) => {
  return (
    <div className="bg-gray-200 rounded-sm w-4/5 max-w-lg flex flex-col gap-4 p-6">
      <SortableContext
        items={tasks.map((task) => ({ ...task, id: task.id.toString() }))}
        strategy={verticalListSortingStrategy}
      >
        {tasks.map((task) => (
          <Task id={task.id} title={task.title} key={task.id.toString()}></Task>
        ))}
      </SortableContext>
    </div>
  );
};
export default Column;
