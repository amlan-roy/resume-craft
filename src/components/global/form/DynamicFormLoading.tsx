import React from "react";
import Loading from "@/components/global/form/form-sections/Loading";

/*
Creating a separate component for loading because of an issue with react hook forms. 
The form data is fetch async, and if once the default data is set in the react hook form method inside dynamic form component, 
it is not updated once the default values data is updated. 
Hence we should not init the method until the data is final. 
Hence we are using a separate loading component to show the loading state. 
Can fix this later, but for now this is the temp solution.
*/
type DynamicFormLoadingProps = {};

const DynamicFormLoading: React.FC<DynamicFormLoadingProps> = () => {
  return (
    <div className="space-y-8">
      <section className="w-full flex flex-col gap-8">
        <Loading />
        <Loading />
      </section>
    </div>
  );
};
export default DynamicFormLoading;
