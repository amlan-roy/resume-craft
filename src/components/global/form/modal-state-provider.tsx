"use client";

import React, { useCallback } from "react";

type ModalStateProviderProps = {
  children: React.ReactNode;
};

type ModalState = {
  title?: string;
  message?: string;
  isOpen: boolean;
  confirmText?: string;
  cancelText?: string;
  onConfirm?: () => void;
  onCancel?: () => void;
};

type ModalStateContextType = {
  modalState: ModalState;
  setModalState: (newState: ModalState) => void;
};

export const ModalStateContext = React.createContext<ModalStateContextType>({
  modalState: {
    isOpen: false,
  },
  setModalState: () => console.warn("No modal state provider"),
});

const ModalStateProvider: React.FC<ModalStateProviderProps> = ({
  children,
}) => {
  const [modalState, setModalState] = React.useState<ModalState>({
    isOpen: false,
  });

  const setModalStateHandler = useCallback((newState: ModalState) => {
    setModalState(() => newState);
  }, []);

  const modalStateContextValue = React.useMemo(() => {
    return {
      modalState,
      setModalState: setModalStateHandler,
    };
  }, [modalState, setModalStateHandler]);

  return (
    <ModalStateContext.Provider value={modalStateContextValue}>
      {children}
    </ModalStateContext.Provider>
  );
};

export default ModalStateProvider;
