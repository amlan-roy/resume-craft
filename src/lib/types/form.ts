import * as z from "zod";

export enum SECTION {
  ADDITIONAL = "ADDITIONAL",
  BASIC_DETAILS = "BASIC_DETAILS",
  EDUCATION = "EDUCATION",
  PROFESSIONAL_SUMMARY = "PROFESSIONAL_SUMMARY",
  PROJECTS = "PROJECTS",
  SKILLS = "SKILLS",
  WORK_EXPERIENCE = "WORK_EXPERIENCE",
}

export const durationFieldSchema = z
  .object({
    startDate: z.date().optional().nullish(),
    endDate: z.date().optional().nullish(),
    current: z.boolean().optional().nullish(),
  })
  .refine(
    (data) => {
      return !(data.endDate || !!data.current) || !!data.startDate;
    },
    (data) => {
      if (data.endDate) {
        return {
          message: "Start Date is required when you select an end date",
          path: ["startDate"],
        };
      }
      return {
        message: "Start Date is required when you select the current checkbox",
        path: ["startDate"],
      };
    }
  );

export type TDuration = {
  startDate?: string | null | undefined;
  endDate?: string | null | undefined;
  current?: boolean | null | undefined;
};

export const basicDetailsFieldSchema = z.object({
  name: z.string().min(1, "Name is required"),
  email: z.string().email().optional().nullish().or(z.literal("")),
  phone: z.string().optional().nullish().or(z.literal("")),
  location: z.string().optional().nullish().or(z.literal("")),
  portfolioUrl: z
    .string()
    .url("Invalid URL. URL should be of the format: 'https://www.abcd.com'")
    .optional()
    .nullish()
    .or(z.literal("")),
  githubUrl: z
    .string()
    .url("Invalid URL. URL should be of the format: 'https://www.abcd.com'")
    .optional()
    .nullish()
    .or(z.literal("")),
  linkedinUrl: z
    .string()
    .url("Invalid URL. URL should be of the format: 'https://www.abcd.com'")
    .optional()
    .nullish()
    .or(z.literal("")),
  stackoverflowUrl: z
    .string()
    .url("Invalid URL. URL should be of the format: 'https://www.abcd.com'")
    .optional()
    .nullish()
    .or(z.literal("")),
  blogUrl: z
    .string()
    .url("Invalid URL. URL should be of the format: 'https://www.abcd.com'")
    .optional()
    .nullish()
    .or(z.literal("")),
});

export const workExperienceFieldSchema = z.object({
  jobTitle: z.string().min(1, "You need to enter the job title"),
  jobSubtitle: z.string().optional().nullish().or(z.literal("")),
  companyName: z.string().optional().nullish().or(z.literal("")),
  location: z.string().optional().nullish().or(z.literal("")),
  duration: durationFieldSchema.optional().nullish(),
  details: z
    .string()
    .min(1, "You need to enter the description of your job experience"),
});

export const projectsFieldSchema = z.object({
  projectTitle: z
    .string()
    .min(1, "You need to enter the title of your project"),
  projectSubtitle: z.string().optional().nullish().or(z.literal("")),
  projectUrl: z.string().optional().nullish().or(z.literal("")),
  associatedWith: z.string().optional().nullish().or(z.literal("")),
  duration: durationFieldSchema.optional().nullish(),
  details: z.string().optional().nullish().or(z.literal("")),
});
export const skillsFieldSchema = z.object({
  title: z.string().optional().nullish().or(z.literal("")),
  skills: z.string().min(1, "This is a required field"),
});

export const educationFieldSchema = z.object({
  universityName: z.string().min(1, "University Name is required"),
  degreeName: z.string().min(1, "Degree Name is required"),
  majorName: z.string().optional().nullish().or(z.literal("")),
  grade: z.string().optional().nullish().or(z.literal("")),
  location: z.string().optional().nullish().or(z.literal("")),
  duration: durationFieldSchema.optional().nullish(),
});

export const basicDetailsSectionSchema = z.object({
  type: z.literal(SECTION.BASIC_DETAILS),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: basicDetailsFieldSchema,
});
export const professionalSummarySectionSchema = z.object({
  type: z.literal(SECTION.PROFESSIONAL_SUMMARY),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: z.object({
    value: z.string().min(1, "You can not leave this field as empty"),
  }),
});
export const workExperienceSectionSchema = z.object({
  type: z.literal(SECTION.WORK_EXPERIENCE),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: z.array(workExperienceFieldSchema),
});
export const projectsSectionSchema = z.object({
  type: z.literal(SECTION.PROJECTS),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: z.array(projectsFieldSchema),
});
export const skillsSectionSchema = z.object({
  type: z.literal(SECTION.SKILLS),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: z.array(skillsFieldSchema),
});
export const educationSectionSchema = z.object({
  type: z.literal(SECTION.EDUCATION),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: z.array(educationFieldSchema),
});
export const additionalSectionSchema = z.object({
  type: z.literal(SECTION.ADDITIONAL),
  sectionTitle: z.string().min(1, "The section title is required"),
  fields: z.object({
    value: z.string().min(1, "This information is required"),
  }),
});

export const sectionTypes = z.union([
  professionalSummarySectionSchema,
  workExperienceSectionSchema,
  projectsSectionSchema,
  skillsSectionSchema,
  educationSectionSchema,
  additionalSectionSchema,
]);

export const formSchema = z.object({
  basicDetails: basicDetailsSectionSchema,
  optionalSections: z.array(sectionTypes),
});

export type formType = z.infer<typeof formSchema>;

export const resumeVariantGenerationFormSchema = z.object({
  jobDescription: z
    .string()
    .min(
      1,
      "It looks like you have missed to add the job description. A job description is required for generating a variant of your resume."
    ),
  modifiedResumeJSON: z
    .string()
    .min(1, "Please enter the JSON response from chat GPT")
    .refine((data) => {
      return formSchema.safeParse(data).success;
    }, "The entered data is not in the correct format. Please try entering both the commands in a new chatGPT conversation. If the issue continues, please write to us"),
  customResumeName: z.string().optional().nullish().or(z.literal("")),
});
