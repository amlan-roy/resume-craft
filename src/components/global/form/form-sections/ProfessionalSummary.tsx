import React from "react";
import { Trash2Icon } from "lucide-react";
import { Card, CardHeader, CardTitle, CardContent } from "@/components/ui/card";
import TextInput from "@/components/global/form/form-inputs/TextInput";
import HiddenInput from "@/components/global/form/form-inputs/HiddenInput";
import { SECTION, formType } from "@/lib/types/form";
import { useFormContext } from "react-hook-form";

type ProfessionalSummaryProps = {
  deleteSection: () => void;
  index?: number;
};

const fieldName = "optionalSections";

const ProfessionalSummary: React.FC<ProfessionalSummaryProps> = ({
  deleteSection,
  index,
}) => {
  const {
    register,
    formState: { errors },
  } = useFormContext<formType>();
  return (
    <Card
      data-testid="form-sections__professional-summary"
      data-card-type={SECTION.PROFESSIONAL_SUMMARY}
    >
      <HiddenInput
        fieldName={
          index !== undefined && index !== null
            ? `${fieldName}.${index}.type`
            : undefined
        }
        value={SECTION.PROFESSIONAL_SUMMARY}
        register={register}
      />
      <CardHeader className="text-brand-neutral-11 flex flex-row flex-wrap w-full justify-between">
        <CardTitle className="w-full max-w-[75%]">
          <TextInput
            fieldName={
              index !== undefined && index !== null
                ? `${fieldName}.${index}.sectionTitle`
                : undefined
            }
            register={register}
            inputClassName="text-xl md:text-2xl py-6"
            placeholder="Section title"
            errorMessage={
              index !== undefined
                ? errors?.[fieldName]?.[index]?.sectionTitle?.message
                : undefined
            }
          />
        </CardTitle>

        <button
          className="ml-auto"
          onClick={deleteSection}
          type="button"
          data-testid="form-sections__delete-icon"
        >
          <Trash2Icon />
        </button>
      </CardHeader>
      <CardContent className="flex flex-wrap w-full gap-5">
        <TextInput
          fieldName={
            index !== undefined && index !== null
              ? `${fieldName}.${index}.fields.value`
              : undefined
          }
          register={register}
          multiline
          className="w-full"
          placeholder="Galactic Theorist with decades of experience, specializing in unraveling the fabric of the cosmos and visiting islands. Proficient in quantum mechanics, general relativity, and little people. Experienced in solving complex quadratic equations."
          errorMessage={
            index !== undefined
              ? // todo: fix this later
                // @ts-expect-error: some ts error
                errors?.[fieldName]?.[index]?.fields?.value?.message
              : undefined
          }
        />
      </CardContent>
    </Card>
  );
};
export default ProfessionalSummary;
