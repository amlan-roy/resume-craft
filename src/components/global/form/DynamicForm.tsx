"use client";

import React, { useContext, useState } from "react";
import { zodResolver } from "@hookform/resolvers/zod";
import { useFieldArray, useForm } from "react-hook-form";
import { z } from "zod";
import {
  DEFAULT_ADDITIONAL_FORM_VALUE,
  DEFAULT_EDUCATION_FORM_VALUE,
  DEFAULT_FORM_VALUE,
  DEFAULT_PROFESSIONAL_SUMMARY_FORM_VALUE,
  DEFAULT_PROJECTS_FORM_VALUE,
  DEFAULT_SKILLS_FORM_VALUE,
  DEFAULT_WORK_EXPERIENCE_FORM_VALUE,
} from "@/lib/const/form/form-data";
import {
  SECTION,
  educationSectionSchema,
  formSchema,
  formType,
  projectsSectionSchema,
  skillsSectionSchema,
  workExperienceSectionSchema,
} from "@/lib/types/form";
import { findFirstFocusable } from "@/lib/utils/findFirstFocusableElemInLastCard";
import { Button } from "@/components/ui/button";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuSeparator,
  DropdownMenuTrigger,
} from "@/components/ui/dropdown-menu";
import { Form } from "@/components/ui/form";
import PopupDialog from "@/components/global/dialog/PopupDialog";
import Additional from "@/components/global/form/form-sections/Additional";
import BasicDetails from "@/components/global/form/form-sections/BasicDetails";
import Education from "@/components/global/form/form-sections/Education";
import Loading from "@/components/global/form/form-sections/Loading";
import ProfessionalSummary from "@/components/global/form/form-sections/ProfessionalSummary";
import Projects from "@/components/global/form/form-sections/Projects";
import Skills from "@/components/global/form/form-sections/Skills";
import WorkExperience from "@/components/global/form/form-sections/WorkExperience";
import ModalStateProvider, { ModalStateContext } from "./modal-state-provider";

type DynamicFormProps = {
  defaultValues?: formType;
  onSubmit: (values: formType) => Promise<void>;
  loading?: boolean;
  confirmationModalTextCopies?: {
    cancelText?: string;
    confirmText?: string;
    description?: string;
    title?: string;
  };
};

const TEXT_COPIES = {
  MODAL: {
    cancelText: "Cancel",
    confirmText: "Confirm",
    description:
      "The existing data (if any) will be overwritten by the new data.",
    title: "Do you want to save this data?",
  },
};

const WrappedDynamicForm: React.FC<DynamicFormProps> = ({ ...props }) => {
  return (
    <ModalStateProvider>
      <DynamicForm {...props} />
    </ModalStateProvider>
  );
};

const DynamicForm: React.FC<DynamicFormProps> = ({
  defaultValues = DEFAULT_FORM_VALUE as formType,
  onSubmit,
  loading,
  confirmationModalTextCopies,
}) => {
  const form = useForm<formType>({
    resolver: zodResolver(formSchema),
    defaultValues: defaultValues,
  });

  const {
    control,
    handleSubmit,
    formState: { errors },
    trigger,
  } = form;

  const {
    fields,
    append,
    remove: deleteSection,
    update: updateSection,
  } = useFieldArray({
    control,
    name: "optionalSections",
  });

  const [focusOnLastSection, setFocusOnLastSection] = useState(false);

  const hasValidationTrigerred = () => {
    return !!Object.keys(form.formState.errors).length;
  };

  // This function is used to trigger validation when a section is deleted
  // Need to manually trigger the validations because once a form field/section is deleted
  // the error message of that deleted field is not cleared and it persists
  // This can cause incorrect error message to be displayed for incorrect fields
  const optionallyTriggerValidationOnDelete = () => {
    if (hasValidationTrigerred()) {
      trigger();
    }
  };

  const addSection = (sectionType: SECTION) => {
    switch (sectionType) {
      case SECTION.PROFESSIONAL_SUMMARY:
        append(DEFAULT_PROFESSIONAL_SUMMARY_FORM_VALUE);
        break;
      case SECTION.WORK_EXPERIENCE:
        append(DEFAULT_WORK_EXPERIENCE_FORM_VALUE);
        break;
      case SECTION.PROJECTS:
        append(DEFAULT_PROJECTS_FORM_VALUE);
        break;
      case SECTION.SKILLS:
        append(DEFAULT_SKILLS_FORM_VALUE);
        break;
      case SECTION.EDUCATION:
        append(DEFAULT_EDUCATION_FORM_VALUE);
        break;
      case SECTION.ADDITIONAL:
        append(DEFAULT_ADDITIONAL_FORM_VALUE);
        break;
    }
  };

  const modalStateContext = useContext(ModalStateContext);

  if (!modalStateContext) {
    throw new Error(
      "DynamicForm component must be used within a ModalStateProvider"
    );
  }

  const {
    isOpen: modalOpen,
    title: modalTitle,
    message: modalMessage,
    confirmText: modalConfirmText,
    cancelText: modalCancelText,
    onConfirm: modalOnConfirm,
    onCancel: modalOnCancel,
  } = modalStateContext.modalState;

  return (
    <ModalStateProvider>
      <Form {...form}>
        <form
          onSubmit={async (e) => {
            e.preventDefault();
            const fieldsValid = await trigger(undefined, { shouldFocus: true });
            fieldsValid &&
              modalStateContext.setModalState({
                isOpen: true,
                title:
                  confirmationModalTextCopies?.title || TEXT_COPIES.MODAL.title,
                message:
                  confirmationModalTextCopies?.description ||
                  TEXT_COPIES.MODAL.description,
                confirmText:
                  confirmationModalTextCopies?.confirmText ||
                  TEXT_COPIES.MODAL.confirmText,
                cancelText:
                  confirmationModalTextCopies?.cancelText ||
                  TEXT_COPIES.MODAL.cancelText,
                onConfirm: () => {
                  handleSubmit(onSubmit)();
                  modalStateContext.setModalState({ isOpen: false });
                },
                onCancel: () => {
                  modalStateContext.setModalState({ isOpen: false });
                },
              });
          }}
          className="space-y-8"
        >
          <section className="w-full flex flex-col gap-8">
            {loading ? (
              <>
                <Loading />
                <Loading />
              </>
            ) : (
              <>
                <BasicDetails fieldName={"basicDetails" as const} />
                {fields.map((field, sectionIndex) => {
                  switch (field.type) {
                    case SECTION.PROFESSIONAL_SUMMARY:
                      return (
                        <ProfessionalSummary
                          key={field.id}
                          deleteSection={() => {
                            deleteSection(sectionIndex);
                            optionallyTriggerValidationOnDelete();
                          }}
                          index={sectionIndex}
                        />
                      );
                    case SECTION.WORK_EXPERIENCE:
                      return (
                        <WorkExperience
                          key={field.id}
                          deleteSection={() => {
                            deleteSection(sectionIndex);
                            optionallyTriggerValidationOnDelete();
                          }}
                          index={sectionIndex}
                          fieldErrors={errors?.optionalSections?.[sectionIndex]}
                          fields={field.fields}
                          updateFields={(
                            addFields?: boolean,
                            index?: number
                          ): void => {
                            if (addFields) {
                              // Directly using form.fields here causes an issue where when we click the add section button after the form is first rendered and if the subsections have some value in it, then
                              // the form.fields will not have the values from the UI. And hence when the button is clicked and a new section is added, then the previous values are lost
                              // However, after this, the values in form.fields are updated correctly and no values are lost on subsequent subsection additions.
                              // So we need to use form.getValues instead
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof workExperienceSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.push({
                                jobTitle: "",
                                details: "",
                                companyName: "",
                                location: "",
                              });
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              return;
                            }
                            if (index || index === 0) {
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof workExperienceSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.splice(index, 1);
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              optionallyTriggerValidationOnDelete();
                            }
                          }}
                        />
                      );
                    case SECTION.PROJECTS:
                      return (
                        <Projects
                          key={field.id}
                          deleteSection={() => {
                            deleteSection(sectionIndex);
                            optionallyTriggerValidationOnDelete();
                          }}
                          index={sectionIndex}
                          fieldErrors={errors?.optionalSections?.[sectionIndex]}
                          fields={field.fields}
                          updateFields={(
                            addFields?: boolean,
                            index?: number
                          ): void => {
                            if (addFields) {
                              // Directly using form.fields here causes an issue where when we click the add section button after the form is first rendered and if the subsections have some value in it, then
                              // the form.fields will not have the values from the UI. And hence when the button is clicked and a new section is added, then the previous values are lost
                              // However, after this, the values in form.fields are updated correctly and no values are lost on subsequent subsection additions.
                              // So we need to use form.getValues instead
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof projectsSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.push({
                                projectTitle: "",
                                details: "",
                              });
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              return;
                            }
                            if (index || index === 0) {
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof projectsSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.splice(index, 1);
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              optionallyTriggerValidationOnDelete();
                            }
                          }}
                        />
                      );
                    case SECTION.SKILLS:
                      return (
                        <Skills
                          key={field.id}
                          deleteSection={() => {
                            deleteSection(sectionIndex);
                            optionallyTriggerValidationOnDelete();
                          }}
                          index={sectionIndex}
                          fieldErrors={errors?.optionalSections?.[sectionIndex]}
                          fields={field.fields}
                          updateFields={(
                            addFields?: boolean,
                            index?: number
                          ): void => {
                            if (addFields) {
                              // Directly using form.fields here causes an issue where when we click the add section button after the form is first rendered and if the subsections have some value in it, then
                              // the form.fields will not have the values from the UI. And hence when the button is clicked and a new section is added, then the previous values are lost
                              // However, after this, the values in form.fields are updated correctly and no values are lost on subsequent subsection additions.
                              // So we need to use form.getValues instead
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof skillsSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.push({
                                skills: "",
                              });
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              return;
                            }
                            if (index || index === 0) {
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof skillsSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.splice(index, 1);
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              optionallyTriggerValidationOnDelete();
                            }
                          }}
                        />
                      );
                    case SECTION.EDUCATION:
                      return (
                        <Education
                          key={field.id}
                          deleteSection={() => {
                            deleteSection(sectionIndex);
                            optionallyTriggerValidationOnDelete();
                          }}
                          index={sectionIndex}
                          fieldErrors={errors?.optionalSections?.[sectionIndex]}
                          fields={field.fields}
                          updateFields={(
                            addFields?: boolean,
                            index?: number
                          ): void => {
                            if (addFields) {
                              // Directly using form.fields here causes an issue where when we click the add section button after the form is first rendered and if the subsections have some value in it, then
                              // the form.fields will not have the values from the UI. And hence when the button is clicked and a new section is added, then the previous values are lost
                              // However, after this, the values in form.fields are updated correctly and no values are lost on subsequent subsection additions.
                              // So we need to use form.getValues instead
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof educationSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.push({
                                universityName: "",
                                degreeName: "",
                              });
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              return;
                            }
                            if (index || index === 0) {
                              const currentField = form.getValues()
                                .optionalSections[sectionIndex] as z.infer<
                                typeof educationSectionSchema
                              >;
                              const currentFields = currentField?.fields;
                              const updatedFields = [...(currentFields || [])];
                              updatedFields.splice(index, 1);
                              updateSection(sectionIndex, {
                                ...currentField,
                                fields: updatedFields,
                              });
                              optionallyTriggerValidationOnDelete();
                            }
                          }}
                        />
                      );
                    case SECTION.ADDITIONAL:
                      return (
                        <Additional
                          key={field.id}
                          deleteSection={() => {
                            deleteSection(sectionIndex);
                            optionallyTriggerValidationOnDelete();
                          }}
                          index={sectionIndex}
                        />
                      );
                  }
                })}
              </>
            )}
          </section>
          {!loading && (
            <>
              <DropdownMenu>
                <DropdownMenuTrigger asChild>
                  <Button variant={"outline"} className="m-4">
                    Add Section
                  </Button>
                </DropdownMenuTrigger>
                <DropdownMenuContent
                  onCloseAutoFocus={(e: Event) => {
                    if (focusOnLastSection) {
                      e.preventDefault();
                      const firstFocusableInLastCard = findFirstFocusable();
                      firstFocusableInLastCard?.scrollIntoView({
                        behavior: "smooth",
                      });
                      firstFocusableInLastCard?.focus();
                      setFocusOnLastSection(false);
                    }
                  }}
                >
                  <DropdownMenuLabel>Select section to add</DropdownMenuLabel>
                  <DropdownMenuSeparator />
                  <DropdownMenuItem
                    data-add-section-menu-item="PROFESSIONAL_SUMMARY"
                    onSelect={() => {
                      setFocusOnLastSection(true);
                      addSection(SECTION.PROFESSIONAL_SUMMARY);
                    }}
                  >
                    Professional Summary
                  </DropdownMenuItem>
                  <DropdownMenuItem
                    data-add-section-menu-item="WORK_EXPERIENCE"
                    onSelect={() => {
                      setFocusOnLastSection(true);
                      addSection(SECTION.WORK_EXPERIENCE);
                    }}
                  >
                    Work Experience
                  </DropdownMenuItem>
                  <DropdownMenuItem
                    data-add-section-menu-item="PROJECTS"
                    onSelect={() => {
                      setFocusOnLastSection(true);
                      addSection(SECTION.PROJECTS);
                    }}
                  >
                    Projects
                  </DropdownMenuItem>
                  <DropdownMenuItem
                    data-add-section-menu-item="SKILLS"
                    onSelect={() => {
                      setFocusOnLastSection(true);
                      addSection(SECTION.SKILLS);
                    }}
                  >
                    Skills
                  </DropdownMenuItem>
                  <DropdownMenuItem
                    data-add-section-menu-item="EDUCATION"
                    onSelect={() => {
                      setFocusOnLastSection(true);
                      addSection(SECTION.EDUCATION);
                    }}
                  >
                    Education
                  </DropdownMenuItem>
                  <DropdownMenuItem
                    data-add-section-menu-item="ADDITIONAL"
                    onSelect={() => {
                      setFocusOnLastSection(true);
                      addSection(SECTION.ADDITIONAL);
                    }}
                  >
                    Additional
                  </DropdownMenuItem>
                </DropdownMenuContent>
              </DropdownMenu>
              <Button type="submit">Submit</Button>
            </>
          )}
        </form>
      </Form>
      <PopupDialog
        open={modalOpen}
        cancelText={modalCancelText || TEXT_COPIES.MODAL.cancelText}
        confirmText={modalConfirmText || TEXT_COPIES.MODAL.confirmText}
        description={modalMessage || TEXT_COPIES.MODAL.description}
        title={modalTitle || TEXT_COPIES.MODAL.title}
        onCancel={() => {
          modalOnCancel?.();
        }}
        onConfirm={() => {
          modalOnConfirm?.();
        }}
      />
    </ModalStateProvider>
  );
};
export default WrappedDynamicForm;
