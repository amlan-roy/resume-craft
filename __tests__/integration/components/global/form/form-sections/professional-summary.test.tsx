import "@testing-library/jest-dom";
import { render, renderHook, screen } from "@testing-library/react";
import { useForm } from "react-hook-form";
import { SECTION } from "@/lib/types/form";
import { Form } from "@/components/ui/form";
import ProfessionalSummary from "@/components/global/form/form-sections/ProfessionalSummary";

describe("integration:components/global/form/form-sections/professional-summary", () => {
  beforeEach(() => {});

  it.skip("snapshot", () => {
    const {
      result: { current },
    } = renderHook(() => useForm());

    const tree = render(
      <Form {...current}>
        <ProfessionalSummary
          sectionTitle="Professional Summary"
          deleteSection={() => {}}
        />
      </Form>
    );

    expect(tree.container).toMatchSnapshot();
  });

  it.skip("ProfessionalSummary component renders correctly", () => {
    render(<ProfessionalSummary deleteSection={() => {}} />);
    const card = screen.getByTestId("form-sections__professional-summary");
    expect(card).toBeInTheDocument();

    const hiddenInput = card.querySelector(
      '[data-testid="hidden-input__container"]'
    );
    expect(hiddenInput).toHaveValue(SECTION.PROFESSIONAL_SUMMARY);

    const deleteButton = screen.getByTestId("form-sections__delete-icon");
    expect(deleteButton).toBeInTheDocument();

    const inputs = screen.getAllByTestId("text-input__container");

    expect(inputs.length).toStrictEqual(2);

    expect(inputs[0].querySelector("label")).not.toBeInTheDocument();
    expect(inputs[0].querySelector("input")).toBeInTheDocument();
    expect(inputs[0].querySelector("p")).not.toBeInTheDocument();

    expect(inputs[1].querySelector("label")).not.toBeInTheDocument();
    expect(inputs[1].querySelector("textarea")).toBeInTheDocument();
    expect(inputs[1].querySelector("p")).not.toBeInTheDocument();
  });

  // todo: Add tests to test error once user events issue is figured out
});
