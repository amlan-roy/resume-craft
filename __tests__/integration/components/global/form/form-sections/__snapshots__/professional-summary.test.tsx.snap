// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration:components/global/form/form-sections/professional-summary snapshot 1`] = `
<div>
  <div
    class="rounded-lg border bg-card text-card-foreground shadow-sm"
    data-card-type="PROFESSIONAL_SUMMARY"
    data-testid="form-sections__professional-summary"
  >
    <input
      class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
      data-testid="hidden-input__container"
      hidden=""
      type="hidden"
      value="PROFESSIONAL_SUMMARY"
    />
    <div
      class="space-y-1.5 p-6 text-brand-neutral-11 flex flex-row flex-wrap w-full justify-between"
    >
      <h3
        class="text-2xl font-semibold leading-none tracking-tight"
      >
        <div
          data-testid="text-input__container"
        >
          <input
            class="flex h-10 w-full rounded-md border border-input bg-background px-3 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 my-3 text-xl md:text-2xl py-6"
            placeholder="Section title"
          />
        </div>
      </h3>
      <button
        class="ml-auto"
        data-testid="form-sections__delete-icon"
        type="button"
      >
        <svg
          class="lucide lucide-trash2 "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 6h18"
          />
          <path
            d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
          />
          <path
            d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
      </button>
    </div>
    <div
      class="p-6 pt-0 flex flex-wrap w-full gap-5"
    >
      <div
        class="w-full"
        data-testid="text-input__container"
      >
        <textarea
          class="flex min-h-[80px] w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 my-3"
          placeholder="Galactic Theorist with decades of experience, specializing in unraveling the fabric of the cosmos and visiting islands. Proficient in quantum mechanics, general relativity, and little people. Experienced in solving complex quadratic equations."
        />
      </div>
    </div>
  </div>
</div>
`;
